engine/common/imagelib/img_png.c

```diff
@@ -0,0 +1,588 @@
+/*
+img_png.c - png format load & save
+Copyright (C) 2019 Andrey Akhmichin
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+*/
+
+#include "miniz.h"
+#include "imagelib.h"
+#include "mathlib.h"
+
+static const char png_sign[] = {0x89, 'P', 'N', 'G', '\r', '\n', 0x1a, '\n'};
+static const char ihdr_sign[] = {'I', 'H', 'D', 'R'};
+static const char idat_sign[] = {'I', 'D', 'A', 'T'};
+static const char iend_sign[] = {'I', 'E', 'N', 'D'};
+
+/*
+=============
+Image_LoadPNG
+=============
+*/
+qboolean Image_LoadPNG( const char *name, const byte *buffer, fs_offset_t filesize )
+{
+	int		ret;
+	short		p, a, b, c, pa, pb, pc;
+	byte		*buf_p, *pixbuf, *raw, *prior, *idat_buf = NULL, *uncompressed_buffer = NULL, *rowend;
+	png_ihdr_t	ihdr_chunk;
+	uint	 	chunk_len, crc32, crc32_check, oldsize = 0, newsize, rowsize;
+	uint	 	uncompressed_size, bpp, pixel_size, i, y, filter_type;
+	qboolean 	has_iend_chunk = false;
+	z_stream 	stream = {0};
+
+	if( filesize < sizeof( png_sign ) )
+		return false;
+
+	buf_p = (byte *)buffer;
+
+	// check png signature
+	if( memcmp( buf_p, png_sign, sizeof( png_sign ) ) )
+		return false;
+
+	// move pointer
+	buf_p += sizeof( png_sign );
+
+	// get IHDR chunk length
+	chunk_len = *buf_p++ << 24;
```

Comment:
This may cause crash on ARM due to unaligned access.

Reply:
A note on my prev comment: reading and writing on ARM is possible only if pointer aligned by 32 bits. So moving pointer by one or two bytes will cause a crash.

I think you can read 32-bits and swap it for big endian, if needed. Engine already have these handy macros: https://github.com/FWGS/xash3d-fwgs/blob/master/common/xash3d_types.h#L88

Reply:
I need swap bytes on little endian, so this macros useless for me.

Reply:
Seems I must use ntohl for little endian,
https://www.w3.org/TR/PNG/#7Integers-and-byte-order

Reply:
It's better to use network conversions here, but keep in mind, that we're still not linking to ws32 on Windows yet directly. So it will fail Windows builds.